### PR TITLE
fix(fswatcher): handle k8s-style symlinks

### LINKED ISSUE #12: Moving certificates doesn't trigger watch
When replacing certificates I'm not writing to an existing file, i've got the new certificates and then i basically copy them into place. In fswatch this is causing a `REMOVE` event and a `CREATED` event, neither of which are handled by certinel https://github.com/cloudflare/certinel/blob/cb169af03858e287e570039166f7ba2ba30629c5/fswatcher/fswatcher.go#L61

The `REMOVE` event is a bit difficult, but i think that at least `CREATED` should be handled here and cause a reload.

### PR DESCRIPTION
**Pull request by terinjokes**
When a Secret is mounted into a container by Kubernetes, the files
created at the "volumeMount" location are symlinks to the real files in
hidden directories. When the kubelet updates the Secret, the new file is
created in a new hidden directory, a hidden symlink is created, which is
then copied over the original symlink.

The previous implementation of fswatcher was watching the original
symlink inode. When the new symlink was copied over, fswatcher observed
a "remove" event, and then saw no other updates. This changeset changes
the behavior to watch the directory, this allows us to observe the new
symlink being copied over the original symlink, and trigger a reload of
certificates.

To preserve behavior when the files are not symlinks, fswatcher checks
for create and modification events for the original file name.

Fixes #12

### CODE REVIEW

File: fswatcher/fswatcher.go [hunk 1]
```diff
@@ -68,9 +73,22 @@ func (w *Sentry) Watch() (certCh <-chan tls.Certificate, errCh <-chan error) {
...
+					if err != nil {
+						w.errChan <- err
+					}
+
+					switch {
+					case !ok:
 						eventsCh = nil
-					} else if event.Op&fsnotify.Write == fsnotify.Write {
+					// 1. The event is for the certificate file and was a creation or modification event
+					case filepath.Clean(event.Name) == certPath && (event.Op&fsnotify.Create != 0 || event.Op&fsnotify.Write != 0),
```
Comment (cbroglie) [inline fswatcher/fswatcher.go:87]: I think it might be easier to follow the logic if you split it into multiple case statements:
```go
// There are 2 separate cases to handle which would trigger a load:
switch {
case !ok:
	eventsCh = nil
case filepath.Clean(event.Name) == certPath && (event.Op&fsnotify.Create != 0 || event.Op&fsnotify.Write != 0):
	// 1. The event is for the certificate file and was a creation or modification event
	w.loadCertificate()
case currentPath != "" && currentPath != realCertPath:
	// 2. We observed the symlink has changed (eg, in Kubernetes)
	realCertPath = currentPath
	w.loadCertificate()
}
```

**Review by cbroglie — DISMISSED** [review]

File: fswatcher/fswatcher.go (hunk 1 above)
Comment (terinjokes) [reply fswatcher/fswatcher.go:87]: This doesn't seem that much easier for me.

Comment (cbroglie) [reply fswatcher/fswatcher.go:87]: I found it hard to follow the case statement split over multiple lines with a comma. Everyone's mileage may vary.

File: fswatcher/fswatcher.go [hunk 2]
```diff
@@ -68,9 +73,22 @@ func (w *Sentry) Watch() (certCh <-chan tls.Certificate, errCh <-chan error) {
 			for eventsCh != nil && errorsCh != nil {
 				select {
 				case event, ok := <-eventsCh:
-					if !ok {
+					// Portions of this case are inspired by spf13/viper's WatchConfig.
+					// (c) 2014 Steve Francia. MIT Licensed.
+					currentPath, err := filepath.EvalSymlinks(certPath)
```
Comment (heckler1) [inline fswatcher/fswatcher.go:78]: This block isn't indented enough it seems like

Comment (heckler1) [reply fswatcher/fswatcher.go:78]: Nevermind, misread this. The diff format confused me

File: fswatcher/fswatcher.go (hunk 1 above)
Comment (heckler1) [reply fswatcher/fswatcher.go:87]: It took me 3 readings to find the comma and interpret it correctly, even after seeing this comment. I think breaking this up into 2 cases makes more sense and is easier to follow.

**Review by heckler1 — APPROVED** [review]

**Review by cbroglie — APPROVED** [review]
